File: typescript/src/providers/geminiProvider.ts
import type { GoogleGenAI } from "@google/genai";
import type { Provider, ProviderRef } from "./types.js";

const SUPPORTED_PREFIXES = ["image/", "audio/", "video/", "application/pdf", "text/"];
const INLINE_SIZE_LIMIT = 20 * 1024 * 1024;

export class GeminiProvider implements Provider {
  constructor(private client: GoogleGenAI) {}

  supportsMediaType(mimeType: string): boolean {
    return SUPPORTED_PREFIXES.some((prefix) => mimeType.startsWith(prefix));
  }

  inlineSizeLimit(): number {
    return INLINE_SIZE_LIMIT;
  }

  async upload(chunks: AsyncIterable<Buffer>, mimeType: string): Promise<ProviderRef> {
    const parts: Buffer[] = [];
    for await (const chunk of chunks) parts.push(chunk);
    const data = Buffer.concat(parts);

    const file = await (this.client as any).files.upload({
      file: new Blob([data], { type: mimeType }),
      config: { mimeType },
    });

    return { id: file.uri, expiresAt: null };
  }

  buildReferenceBlock(ref: ProviderRef, mimeType: string): Record<string, unknown> {
    return { file_data: { mime_type: mimeType, file_uri: ref.id } };
  }

  buildInlineBlock(base64Data: string, mimeType: string): Record<string, unknown> {
    return { inline_data: { mime_type: mimeType, data: base64Data } };
  }
}

File: python/src/siphon/providers/gemini_provider.py
"""Gemini provider adapter: Files API (resumable upload) + generateContent blocks."""

from __future__ import annotations

import io
from typing import AsyncIterator, Iterator

from siphon.providers import ProviderRef

_SUPPORTED_PREFIXES = ("image/", "audio/", "video/", "application/pdf", "text/")
_INLINE_SIZE_LIMIT = 20 * 1024 * 1024  # Gemini's documented inline request-size guidance


class GeminiProvider:
    def __init__(self, client) -> None:
        self._client = client

    def supports_media_type(self, mime_type: str) -> bool:
        return mime_type.startswith(_SUPPORTED_PREFIXES)

    def inline_size_limit(self) -> int:
        return _INLINE_SIZE_LIMIT

    def upload(self, chunks: Iterator[bytes], mime_type: str) -> ProviderRef:
        buffer = io.BytesIO(b"".join(chunks))
        file_obj = self._client.files.upload(
            file=buffer,
            config={"mime_type": mime_type},
        )
        return ProviderRef(id=file_obj.uri, expires_at=None)

    async def upload_async(
        self, chunks: AsyncIterator[bytes] | Iterator[bytes], mime_type: str
    ) -> ProviderRef:
        if hasattr(chunks, "__anext__"):
            data = b"".join([c async for c in chunks])  # type: ignore[union-attr]
        else:
            data = b"".join(chunks)  # type: ignore[arg-type]
        buffer = io.BytesIO(data)
        file_obj = await self._client.aio.files.upload(
            file=buffer,
            config={"mime_type": mime_type},
        )
        return ProviderRef(id=file_obj.uri, expires_at=None)

    def build_reference_block(self, ref: ProviderRef, mime_type: str) -> dict:
        return {"file_data": {"mime_type": mime_type, "file_uri": ref.id}}

    def build_inline_block(self, base64_data: str, mime_type: str) -> dict:
        return {"inline_data": {"mime_type": mime_type, "data": base64_data}}
